pkg/azurecluster/builder.go

```diff
@@ -0,0 +1,68 @@
+package azurecluster
+
+import (
+	"encoding/json"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	capzv1alpha3 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/api/v1alpha3"
+)
+
+type BuilderOption func(azureCluster *capzv1alpha3.AzureCluster) *capzv1alpha3.AzureCluster
+
+func Location(location string) BuilderOption {
+	return func(azureCluster *capzv1alpha3.AzureCluster) *capzv1alpha3.AzureCluster {
+		azureCluster.Spec.Location = location
+		return azureCluster
+	}
+}
+
+func ControlPlaneEndpoint(controlPlaneEndpointHost string, controlPlaneEndpointPort int32) BuilderOption {
+	return func(azureCluster *capzv1alpha3.AzureCluster) *capzv1alpha3.AzureCluster {
+		azureCluster.Spec.ControlPlaneEndpoint.Host = controlPlaneEndpointHost
+		azureCluster.Spec.ControlPlaneEndpoint.Port = controlPlaneEndpointPort
+		return azureCluster
+	}
+}
+
+func BuildAzureCluster(clusterName string, opts ...BuilderOption) *capzv1alpha3.AzureCluster {
```

Comment:
Proposal: In order to avoid unintentional conflicts between tests, would it make sense to change this slightly towards randomized CR naming:
```suggestion
func BuildAzureCluster(opts ...BuilderOption) *capzv1alpha3.AzureCluster {
```

Here, the sane default would be random string generated on function all and if test wants to specify the object name, it could be done with a `BuilderOption()`.

Reply:
I like it!